3단계 - 로또(자동)

sangik-kim opened the pull request:
로또(자동)을 구현했습니다. 처음에 대부분 도메인을 TDD로 구현하였으나 Map을 사용한 부분을 List Object로 변경함에 따라 전체적 리팩토링이 필요해졌습니다. 리팩토링 단계에서 TDD는 쉽지 않았습니다. 아직 연습이 많이 필요한것 같네요.

도메인 모델의 경우 Getter를 사용안하는게 좋다고 들었는데, Getter 없이 DTO로 어떻게 매핑해야될지 고민하다 답이 안나와 몇몇 도메인은 Getter를 추가했습니다. Getter를 쓰지 않고 DTO로 매핑을 어떻게 할 수 있을까요?

생각보다 너무 많은 시간을 이번 단계에서 쓰게 되었네요. 부족함을 많이 느낍니다.. 😢

sangikk committed:
refactor(StringAddCalculatorTest): divide the test method with two assertion into two different test methods

sangikk committed:
refactor(StringAddCalculator): make the code more meaningful

- extract method
- make Pattern as a constant

sangikk committed:
refactor: reorganize the package structure

sangikk committed:
docs(README): add README docs

sangikk committed:
feat(LottoNumber): add a lotto number class

sangikk committed:
feature(Lotto): add a lotto domain class

sangikk committed:
docs(README): add a new requirement that was missing

sangikk committed:
feat(LottoGenerator): add an automatic lotto number generator

sangikk committed:
feat(LottoMachine): add a feature to issue lottos depending on the given money

sangikk committed:
feat(LottoDraw): enable to generate a winning number

sangikk committed:
docs(README): add two more requirements

sangikk committed:
feat(sort): add a sorter to the lotto generator in order to sort the lotto numbers

sangikk committed:
feat(LottoNumbers): check duplicates

sangikk committed:
feat(Division): add a division enum to find a division depending on the match count

sangikk committed:
refactor(test): change class modifiers to package and assertJ usage

- class modifier from public to package for test class
- change isEqualTo to hasToString

sangikk committed:
refactor(LottoNumber): change property name

sangikk committed:
feat(DrawResult): enable to get the lotto draw result

sangikk committed:
feat(DrawResult): add a feature to calculate the earnings rate (with 2 decimal points)

sangikk committed:
refactor(packages): reorganize package structure

sangikk committed:
refactor(file): remove empty file

sangikk committed:
feat(view): add views and dtos

- add Printer and Scanner for console input / output

sangikk committed:
refactor(domain): make the division result data to first class collections

First Class Collections
- add DivisionResult
- add DivisionResults
- add Lottos

Others
- add LottoPurchaseHistory to manage lotto list and its price together
- modify all related DTOs

sangikk committed:
feat(controller): add a controller and main

sangikk committed:
fix(view): fix carriage return symbols

sangikk committed:
fix(exception): modify exception handlers

DTOs
- check if purchase price is positive
- check if the winning number are comma split 6 numbers

Domains
- throw exceptions when lotto cannot be purchased

ConsolePrinter
- add error method to print error message

jaeyeonling commented on src/main/java/study/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,24 @@
+        this.lottoNumbers = lottoNumbers;
+    }
+
+    public Division match(LottoNumbers lottoNumbers) {
```
LottoNumbers와 Lotto의 차이는 해당 메서드의 역할을 하는 것 일까요?
어떠한 의도로 설계해주셨는지 의견이 궁금합니다!

sangik-kim replied:
`LottoNumbers` 는 `LottoNumber`의 리스트만으로 구성된 일급컬랙션입니다.

제가 처음 설계할 때 아래의 요구사항을 도출했습니다.
```
로또는 로또번호를 가지고 있다. 로또번호는 6개의 숫자로 이루어지며, 중복된 숫자는 가질 수 없다.
```
이를 위한 검증 로직을 `LottoNumbers`에 위임함하고 `LottoNumbers`는 단순히 또 다른 `LottoNumbers`와 같은 숫자만 찾는다면, `Lotto`는 추첨 결과 로또번호를 받아 등수를 반환하는 기능을 제공합니다. 지금 다시보니 `match`라는 메소드 이름이 의미를 정확히 전달하지 못한것 같아 수정되어야 할 것 같습니다!

jaeyeonling replied:
`LottoNumbers`가 해당 역할만 가진다면 너무 역할이 적어보이기도 하고, Lotto와 중복될 수 있을 것 같은데요!
설계에 따라 다르겠지만 꼭 `LottoNumbers`와 `Lotto`를 분리해야 할까? 라는 고민도 해보면 좋겠어요 ㅎㅎ

sangik-kim replied:
저는 `LottoNumbers`는 일급컬렉션으로 데이터 검증 및 데이터와 관련된 기능(비교)만 가져야한다고 생각했습니다. 반면에 `Lotto`는 비지니스 (로또)에 있어서 가장 중요한 도메인중 하나이며, 단순 비교가 아니라 주어진 당첨번호에 따른 등수를 반환할 수 있는 비지니스 로직을 가지고 있기에 별도의 도메인이 되어야 한다고 생각했습니다.

하지만 말씀하신대로 불필요한 클래스가 늘어날 수 있기 때문에, 비슷한 클래스 분리는 고민을 더 해봐야 할 것 같네요.

좋은 인사이트 감사 드립니다 👍

jaeyeonling commented on src/main/java/study/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,67 @@
+    static final int LOTTO_NUMBER_SIZE = 6;
+    static final String TO_STRING_DELIMITER = ", ";
+
+    private final List<LottoNumber> lottoNumberList;
```
List와 같은 자료형을 변수명에 사용하면 어떠한 단점이 있을지 고민해보아요

sangik-kim replied:
네 감사합니다. 이 부분도 고민되는 부분인데, 사실 처음에 리스트 명이 `lottoNumbers` 였습니다. 근데 클래스명이랑 똑같은걸 lint에서 발견되어 다른 적당한 이름이 생각나지 않아 List로 붙였는데, 이게 자료구조가 변경되었을때 문제가 되겠군요. 보통 이런 부분(일급 콜랙션)은 어떻게 명명하시나요?

jaeyeonling replied:
일급콜렉션의 경우 콜렉션을 포장하여 하나의 값만 존재하기 때문에 value 혹은 values를 많이 사용하는 편 입니다 :)

sangik-kim replied:
명확한 답변 감사합니다! 👍

jaeyeonling commented on src/main/java/study/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,67 @@
+        return Objects.hash(lottoNumberList);
+    }
+
+    public List<Integer> match(LottoNumbers lottoNumbers) {
```
우선순위가 낮은 equals, hashCode, toString을 클래스 맨 아래에 위치시키면 가독성이 상승할 것 같습니다!

sangik-kim replied:
맞습니다! 실수가 있었네요.

저는 보통 실무에서 equals / hashCode / toString을 거의 Lombok을 이용해서 어노테이션만 추가하는데 미션에선 Lombok을 사용하지 않아서 IDE 기능으로 추가하고 순서를 확인 안하것 같습니다.

jaeyeonling commented on src/main/java/study/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,34 @@
+public class Lottos {
+    private final List<Lotto> lottoList;
+
+    public Lottos(List<Lotto> lottoList) {
+        this.lottoList = lottoList;
+    }
+
+    public List<Lotto> getLottoList() {
+        return lottoList;
+    }
```
list 자체를 생성자로 받거나 외부로 공개하게 된다면 외부에서 해당 값을 악용할 수 있는데요.
외부에서 해당 값을 조작할 수 없도록 제한할 수 있는 방법이 없을지 고민해보아요.

sangik-kim replied:
객체 내부의 리스트와 같은 자료구조가 밖으로 노출된다면 setter 없이 데이터 변경이 가능하겠군요.

이 부분이 추가된 이유는 View에서 표현되기 위해 DTO로 변환하는 과정에서 Lotto데이터가 필요해서였습니다. 
혹시, 이 부분을 로또 번호를 가지고 List<Integer>를 새로 생성해서 반환하는건 어떨까요? 외부에서 수정한다고 해도 객체내부 데이터가 변경되진 않을것 같습니다. 다른 방법으로는 로또 번호를 스트링으로 반환할 수 있을것 같습니다.

아직 생성자에 List로 받았을 때, 어떻게 악용 할 수 있는지 짐작이 안가는데 혹시 답변 가능하신가요?

jaeyeonling replied:
생성자로 List를 받았을 때도 동일하게 새로운 List를 생성할 수 있을 것 같습니다~
관련 키워드로 `방어적 복사` 혹은 `불변 컬렉션`에 대해 학습해보면 좋을 것 같아요 :)

sangik-kim replied:
아! 그렇군요. 사용자가 입력한 리스트 그대로 객체를 생성하게 되면 사용자는 생성된 객체에 setter가 없어도 생성할 때 따로 참조값을 빼두었다면 내부 값 수정이 가능하겠네요! 생각지도 못한 부분인데 감사합니다 🙇

jaeyeonling commented on src/main/java/study/lotto/domain/draw/DivisionResults.java:
```diff
@@ -0,0 +1,56 @@
+        return divisionResultList;
+    }
+
+    public BigDecimal getAllPrize() {
+        return divisionResultList.stream()
+                .map(DivisionResult::calculatePrize)
+                .reduce(BigDecimal.ZERO, (prizeTotal1, prizeTotal2) -> prizeTotal1.add(prizeTotal2));
+    }
+
+    private DivisionResult findDivisionResult(Division division, List<DivisionResult> divisionResultList) {
+        return divisionResultList.stream().filter(divisionResult -> divisionResult.isDivisionSame(division)).findFirst()
+                .orElse(new DivisionResult(division));
+    }
```
stream() 다음에 줄바꿈을 하는 경우도 있고, 이어서 작성하는 경우도 있는데 규칙이 있을까요?
규칙이 없다면 스타일을 통일해주시면 좋을 것 같은데요.
더 가독성이 좋은 방향으로 통일하면 좋을 것 같네요 ㅎㅎ

sangik-kim replied:
저는 보통 한줄에 작성이 가능하면 한줄에 쓰고, 아니면 stream()에서 개행 합니다. 이번에 정신없이 리팩토링하다 엉망으로 작성했네요. 코드 리뷰를 올리기전에 스스로 리뷰하는 습관을 가져야 할 것 같습니다 😢

jaeyeonling commented on src/main/java/study/lotto/domain/lottomachine/LottoPrice.java:
```diff
@@ -0,0 +1,57 @@
+        return money.divideAndRemainder(price)[0].intValue();
+    }
+
+    public BigDecimal total(int numberOfLotto) {
```
해당 메서드명으로 해당 메서드가 하는 역할을 잘 나타내고 있는지 고민해보면 좋겠어요!

sangik-kim replied:
코딩을 하면서 가장 어려운 부분중 하나가 명명하는거라고 생각하고 실제로 생각보다 많은 시간을 변수명 고민으로 사용합니다. 이 메소드는 구매한 로또 수에 따른 전체 구입 금액을 제공하기 위한 메소드입니다. 지금 다시보니, `total` 보다는 `totalPrice`가 적합해 보입니다.

jaeyeonling commented on src/main/java/study/lotto/domain/sorter/IntegerSorter.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.List;
+
+public interface IntegerSorter {
```
해당 인터페이스를 분리해주신 이유가 있을까요?

sangik-kim replied:
로또 번호 생성기에서 사용하고 있는 인터페이스인데, 혹시나 번호 정렬의 요구사항이 변경됨에 유연하게 대처하기 위해 별도의 인터페이스로 두었습니다. 이 인터페이스를 구현 함으로서 역순이나 무작위로 번호 순서를 지정할 수 있습니다.

jaeyeonling replied:
개인적으론 View의 요구사항에 가까워 보이기 때문에 정렬에 대한 부분이 필요할까? 라는 의문이 있는데요
상익님은 해당 번호의 정렬이 도메인의 역할로 보신걸까요?

sangik-kim replied:
View에서 처리해야할 요구사항으로 볼 수 도 있겠네요. 저는 로또 번호 순서 자체도 비지니스 도메인으로 판단했고 View에서 처리한다고 생각하질 못했네요. 그래서 [docs/README.md](https://github.com/next-step/java-lotto-pro/pull/451/files#diff-03722b0c7d0a72278186f573989eb08fb554e93217744e19d036200ee1512532)에 하나의 요구사항으로 빼뒀습니다.

재연님 의견을 보고나니 View에서 처리해도 전혀 문제될 건 없어 보입니다. 👍

jaeyeonling commented on src/main/java/study/lotto/domain/lottomachine/LottoMachine.java:
```diff
@@ -0,0 +1,43 @@
+
+    private LottoPurchaseHistory issue(BigDecimal money) {
+        int count = numberOfLottos(money);
+        List<Lotto> lottoList = IntStream.range(0, count).mapToObj(i -> generateLotto()).collect(Collectors.toList());
```
```suggestion
        List<Lotto> lottoList = Stream.generate(this::generateLotto).limit(count).collect(Collectors.toList());
```

와 같이도 사용할 수 있겠네요~

sangik-kim replied:
네 제안해 주신 방법으로 사용한다면 불필요한 변수 i 없이도 가능하고 가독성에도 더 도움이 될 것 같습니다. 👍

jaeyeonling commented on src/main/java/study/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,34 @@
+        this.lottoList = lottoList;
+    }
+
+    public List<Lotto> getLottoList() {
```
Q. 도메인 모델의 경우 Getter를 사용안하는게 좋다고 들었는데, Getter 없이 DTO로 어떻게 매핑해야될지 고민하다 답이 안나와 몇몇 도메인은 Getter를 추가했습니다. Getter를 쓰지 않고 DTO로 매핑을 어떻게 할 수 있을까요?
A. 도메인에서 Getter를 사용하지 말라는 뜻은 도메인끼리 메시지를 던지는 방식으로 구현하라! 는 의미입니다!
실제 View를 위해 getter를 생성해주는 것은 괜찮습니다 😄

sangik-kim replied:
감사합니다! 👍

jaeyeonling reviewed (changes requested):
로또를 잘 구현해주셨어요 👍 
질문의 답변과 코멘트 남겨두었으니 확인 부탁드립니다 😄

sangikk committed:
refactor(domain): refactor codes based on the comments of the reviewer

Refactoring
- rename method name and variable names to improve readability (Lotto, Lottos,  LottoPrice)
- move methods based on the priority of criticality
- change the usage of Stream to improve readability
- change Lottos not to directly return its List to prevent abusing it

sangikk committed:
refactor: re-format usage of Java stream

sangikk committed:
refactor: refactor first class collections and apply defensive copy

Change names of instance variables
- change names to value

sangik-kim commented:
남겨주신 리뷰를 바탕으로 수정하였습니다! 감사합니다 🙇

jaeyeonling commented on src/main/java/study/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,28 @@
+public class Lotto {
+    private final LottoNumbers lottoNumbers;
+
+    public Lotto(Lotto lotto) {
```
Lotto로 Lotto를 생성하는 경우는 어떤 케이스가 있을까요?

sangik-kim replied:
Lottos에서 List<Lotto>를 생성자에서 받는데, `방어적 복사`를 할 때, Lotto 생성자를 이용해 같은 값을 가진 객체를 생성합니다.

jaeyeonling commented on src/main/java/study/lotto/domain/LottoNumbers.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    public List<Integer> numbers() {
+        return value.stream().map(LottoNumber::toString).map(Integer::valueOf).collect(Collectors.toList());
```
```suggestion
        return value.stream()
                .map(LottoNumber::toString)
                .map(Integer::valueOf)
                .collect(Collectors.toList());
```

스트림 사용 시 라인을 나눠서 작성하면 가독성이 상승합니다!

sangik-kim replied:
네 알겠습니다. 👍  전 보통 한줄에 끝나는 스트림은 그대로 두고 한줄 이상일 경우 stream 다음부터 개행했는데, 한 줄 상관 없이 개행하도록 작성해야겠네요.

jaeyeonling replied:
한 줄 단위로 개행한다면 intelliJ의 힌트도 확인하실 수 있답니다 😄 

![image](https://user-images.githubusercontent.com/40811999/168717206-5f649272-e097-46d7-af8e-299ddecdce17.png)

jaeyeonling commented on src/main/java/study/lotto/domain/draw/DivisionResults.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public List<DivisionResult> get() {
+        return value;
```
https://github.com/next-step/java-lotto-pro/pull/451#discussion_r872993840 해당 코멘트를 참고해주세요!

sangik-kim replied:
이 부분 빼먹었네요.. 😞

jaeyeonling commented on src/main/java/study/lotto/domain/lottomachine/LottoPurchaseHistory.java:
```diff
@@ -0,0 +1,28 @@
+    private final Lottos lottos;
+    private final BigDecimal totalPrice;
+
+    public LottoPurchaseHistory(List<Lotto> lottoList, BigDecimal totalPrice) {
```
https://github.com/next-step/java-lotto-pro/pull/451#discussion_r872993624 해당 코멘트를 확인해주세요!

sangik-kim replied:
[Lottos 불변 컬렉션 생성](https://github.com/next-step/java-lotto-pro/pull/451/files#diff-809f1cf7a81f684d0f8fb2191ecf2a9af00adc75b7b3ddb3a257166908308c16R14-R16)

이 부분은 Lottos 객체를 생성할 때, 불변으로 처리 됩니다 😄

jaeyeonling replied:
`List<Lotto> lottoList`에서 List와 같은 자료형을 변수명에 담는다면 어떠한 문제가 있을지에 대한 코멘트였습니다~

jaeyeonling reviewed (approved):
코멘트를 잘 반영해주셨습니다 👍 
역할에 맞게 객체를 분리하는 것은 좋지만 과한 설계를 할 경우의 단점도 존재하는데요!
다음 단계에선 최소한의 설계로 리팩토링하면 좋을 것 같습니다 :) 
추가로 소소한 코멘트 남겨두었으니 확인 부탁드리고 다음 단계 진행해주세요~
